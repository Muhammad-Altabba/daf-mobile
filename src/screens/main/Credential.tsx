import React, { useEffect, useState } from 'react'
import {
  Container,
  Screen,
  Credential,
  Device,
  Text,
  RadioBtn,
  Constants,
  Typings,
} from '@kancha/kancha-ui'
import { NavigationStackScreenProps } from 'react-navigation-stack'
import { SharedElement } from 'react-navigation-shared-element'
import { HeaderButtons, Item } from 'react-navigation-header-buttons'
import { FlatList } from 'react-native'
import { Colors } from '../../theme'
import { ScrollView } from 'react-native-gesture-handler'

interface CredentialStyle {
  background: Typings.BrandPropOptions
  shadow: number
}

interface Props extends NavigationStackScreenProps {}

const CredentialDetail: React.FC<Props> & { sharedElements: any } & {
  navigationOptions: any
} = ({ navigation }) => {
  const credentials = navigation.getParam('credentials')
  const initialCredential = navigation.getParam('credentialIndex', 0)
  const [sharingMode, toggleSharingMode] = useState(false)
  const [selected, updateSelected] = useState<number[]>([])

  const isSelected = (index: number) => {
    return selected.includes(index)
  }

  const selectedStyle = (index: number): CredentialStyle => {
    return {
      background: isSelected(index) ? 'primary' : 'secondary',
      shadow: 0,
    }
  }

  const selectCredential = (index: number) => {
    if (selected.includes(index)) {
      updateSelected(selected.filter((item: number) => item !== index))
    } else {
      updateSelected(selected.concat([index]))
    }
  }

  useEffect(() => {
    navigation.setParams({ sharingMode, toggleSharingMode })
  }, [sharingMode])

  return (
    // <Screen scrollEnabled background={'primary'}>
    <Container flex={1} backgroundColor={Colors.BLACK}>
      {sharingMode && (
        <Container padding>
          <Text type={Constants.TextTypes.Body} textColor={Colors.LIGHT_GREY}>
            Select credential(s) for sharing. Flow to be completed.
          </Text>
        </Container>
      )}
      <FlatList
        contentOffset={{
          x: (Device.width - 15) * initialCredential,
          y: 0,
        }}
        horizontal
        pagingEnabled
        showsHorizontalScrollIndicator={false}
        snapToAlignment={'center'}
        keyExtractor={(item: any) => item.hash}
        data={credentials}
        renderItem={({ item, index }: any) => {
          return (
            <Container w={Device.width - 10} padding paddingRight={10}>
              <ScrollView scrollEventThrottle={16}>
                <SharedElement id={item.hash + item.rowId}>
                  <Credential
                    onPress={() => sharingMode && selectCredential(index)}
                    detailMode
                    jwt={item.jwt}
                    issuer={item.iss}
                    subject={item.sub}
                    fields={item.fields}
                    exp={item.exp}
                    {...selectedStyle(index)}
                  />
                  {sharingMode && (
                    <RadioBtn
                      selected={isSelected(index)}
                      onPress={() => selectCredential(index)}
                    >
                      {isSelected(index) && 'Sharing'}
                    </RadioBtn>
                  )}
                </SharedElement>
              </ScrollView>
            </Container>
          )
        }}
      ></FlatList>
      {/* <Container flexDirection={'row'} justifyContent={'center'}>
        {credentials.map((dot: any) => {
          return (
            <Container
              key={dot.hash}
              br={5}
              w={10}
              h={10}
              backgroundColor={Colors.MEDIUM_GREY}
              margin={4}
            />
          )
        })}
      </Container> */}
    </Container>
    // </Screen>
  )
}

CredentialDetail.navigationOptions = ({ navigation }: any) => {
  const { sharingMode, toggleSharingMode } = navigation.state.params || {}

  return {
    title: 'Credential',
    headerStyle: {
      backgroundColor: '#f4511e',
    },
    headerLeft: (
      <HeaderButtons>
        <Item
          title={'Done'}
          onPress={navigation.dismiss}
          color={Colors.WHITE}
        />
      </HeaderButtons>
    ),
    headerRight: (
      <HeaderButtons>
        {sharingMode ? (
          <Item
            title={'Cancel'}
            onPress={() => toggleSharingMode(false)}
            color={Colors.WHITE}
          />
        ) : (
          <Item
            title={'Share'}
            onPress={() => toggleSharingMode(true)}
            color={Colors.WHITE}
          />
        )}
      </HeaderButtons>
    ),
  }
}

CredentialDetail.sharedElements = (navigation: any) => {
  const credentials = navigation.getParam('credentials')
  return credentials.map((item: any) => {
    return { id: item.hash + item.rowId, animation: 'fade', resize: 'clip' }
  })
}

export default CredentialDetail
