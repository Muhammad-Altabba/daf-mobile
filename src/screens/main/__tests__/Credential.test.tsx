import 'react-native'
import React from 'react'
import Credential from '../Credential'
import { render } from 'react-native-testing-library'

const sertoVerifiableCredential = {
  iss: 'Serto Identity Platform',
  sub: 'did:ethr:0xf3beac30c498d9e26865f34fcaa57dbb935b0d74',
  iat: 1562769371,
  exp: 1579478400,
  claim: {
    'Serto ID': {
      name: 'Sarah Adamson',
      dateOfBirth: '22-01-75',
      country: 'USA',
      children: [
        {
          name: 'Bob',
          age: 4,
        },
        {
          name: 'Alice',
          age: 9,
        },
      ],
    },
  },
  vc: [],
}

describe('Credential', () => {
  it('renders correctly with data', () => {
    const navigation = {
      navigate: jest.fn(),
      getParam: jest
        .fn()
        .mockReturnValue({ ...sertoVerifiableCredential, type: 'Some VC' }),
    }
    //@ts-ignore
    const tree = render(<Credential navigation={navigation} />).toJSON()

    expect(navigation.getParam).toBeCalled()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly without a type', () => {
    const navigation = {
      navigate: jest.fn(),
      getParam: jest.fn().mockReturnValue(sertoVerifiableCredential),
    }
    //@ts-ignore
    const tree = render(<Credential navigation={navigation} />).toJSON()

    expect(navigation.getParam).toBeCalled()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly without data', () => {
    const navigation = {
      navigate: jest.fn(),
      getParam: jest.fn().mockReturnValue(null),
    }
    //@ts-ignore
    const tree = render(<Credential navigation={navigation} />).toJSON()

    expect(navigation.getParam).toBeCalled()
    expect(tree).toMatchSnapshot()
  })
})
