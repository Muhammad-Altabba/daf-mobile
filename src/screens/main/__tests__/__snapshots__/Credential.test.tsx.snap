// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Credential renders correctly without data 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "alignSelf": "stretch",
        "backgroundColor": "#ffffff",
        "borderRadius": undefined,
        "borderTopLeftRadius": 10,
        "borderTopRightRadius": 10,
        "flex": 1,
        "flexDirection": undefined,
        "height": undefined,
        "justifyContent": undefined,
        "margin": undefined,
        "marginBottom": undefined,
        "marginLeft": 10,
        "marginRight": 10,
        "marginTop": 50,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "shadowColor": "#000000",
        "shadowOpacity": 0.2,
        "shadowRadius": 20,
        "width": undefined,
      }
    }
  >
    <View
      pointerEvents="auto"
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": undefined,
          "borderRadius": undefined,
          "flex": undefined,
          "flexDirection": undefined,
          "height": undefined,
          "justifyContent": "center",
          "margin": undefined,
          "marginBottom": undefined,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": undefined,
          "padding": 8,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "width": undefined,
        }
      }
    >
      <View
        pointerEvents="auto"
        style={
          Object {
            "alignItems": undefined,
            "backgroundColor": "#BBBBBB",
            "borderRadius": 5,
            "flex": undefined,
            "flexDirection": undefined,
            "height": 5,
            "justifyContent": undefined,
            "margin": undefined,
            "marginBottom": undefined,
            "marginLeft": undefined,
            "marginRight": undefined,
            "marginTop": undefined,
            "padding": undefined,
            "paddingBottom": undefined,
            "paddingHorizontal": undefined,
            "paddingLeft": undefined,
            "paddingRight": undefined,
            "paddingTop": undefined,
            "width": 50,
          }
        }
      />
    </View>
  </View>
</View>
`;
